File: aoc15/ocaml/day_04/day_04.ml
let input = Input.read "input" in
let hash_is_valid prefix n =
  let hash = Digest.to_hex (Digest.string (input ^ string_of_int n)) in
  String.starts_with ~prefix hash
in

let first = Seq.find (hash_is_valid "00000") (Seq.ints 1) in
let second = Seq.find (hash_is_valid "000000") (Seq.ints 1) in
Printf.printf "First: %d\n" (Option.get first);
Printf.printf "Second: %d\n" (Option.get second)

File: aoc15/go/day_04/day_04.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func find_key(key, prefix string) int {
	for i := 1; ; i++ {
		hash := md5.Sum([]byte(key + strconv.Itoa(i)))
		str := hex.EncodeToString(hash[:])
		if strings.HasPrefix(str, prefix) {
			return i
		}
	}
	return 0
}

func main() {
	input, _ := os.ReadFile("input")
	key := strings.TrimSpace(string(input))
	fmt.Println("First:", find_key(key, "00000"))
	fmt.Println("Second:", find_key(key, "000000"))
}
